### ISSUE TITLE
Is there a way to ignore certain files/folders from the tests?

### ISSUE THREAD

**Issue by danielsitnik** [ts-arch/ts-arch#49]
Hello, I'm looking for a way to ignore certain files or folders from the tests. I could not find anything on the documentation or other issues about this topic.

Given my folder structure, TSArch is trying to enforce its rules on the test files, which is not ideal.
I'll give some context and two problems I'm facing because of this.

Thanks in advance for any help and TSArch is awesome. 😄

### Context
In my project I have the following folder structure:

```
root
├── tsconfig.json
└── src
    ├── application
    │   └── service
    │       ├── MyService.ts
    │       └── WrongName.ts
    └── test
        └── application
            └── service
                └── MyService.test.ts
```

The `tsconfig.json` file is at the root folder.
The source code is at the `src` folder (also root), where application code goes into `src/application` and tests go into `src/test`.

There are some sub-folders under `src/application`, so "service" classes go under `src/application/service`, as is the case with `MyService.ts` on this example.

The test folder follows the application folder structure. So a test for the class mentioned above goes into `src/test/application/service` and its name ends with `.test.ts`.

### Problem 1
I have some test cases like this one to validate the file name suffix:
```typescript
test('service classes have Service suffix', async () => {
  const rule = filesOfProject()
    .inFolder('application/service')
    .should()
    .matchPattern('.*Service.ts');

  await expect(rule).toPassAsync();
});
```

The problem with this test is that it also picks up the files under `src/test/application/service`,  so it fails for the file `WrongName.ts`as expected, but also fails for `MyService.test.ts`.

I have tried to change it to be more specific like `.inFolder('src/application/service')`, but this gives me a false positive: the test passes without picking up the `WrongName.ts` file.

Right now the workaround has been to use a rule that also accepts files ending in .test.ts: `.matchPattern('.*Service(.ts|.test.ts)')`

### Problem 2
I also have a test to check for violations in application layers, which is defined like this:
```typescript
test('application adheres to architecture layers', async () => {
  const diagram = path.resolve(__dirname, 'architecture.puml');
  const rule = await slicesOfProject()
    .definedBy('src/(**)/')
    .should()
    .adhereToDiagramInFile(diagram);

  await expect(rule).toPassAsync();
});
```
As you see my slices are defined like `src/(**)/`, so it ends up considering `src/test/` as a valid slice.
This test was also failing because files in the test slice coud not depend on files from the application slice.

So in this case I had to change my architecture.puml to include something like this:
```
[test] --> [application]
```

**Comment by torfmaster** [comment]
Thank you for taking your time creating the issue. For me it seems that the issues are implied by a missing distinction between test scope and production scope in typescript (as it is usually the case for Java projects, for example). I see two solutions:

* do it the nx-way: have separate `tsconfig.json` files for production, where the production version ignores the test scope (using `exclude` statements). If that would be possible, you could invoke `slicesOfProject(''tsconfig.prod.conf)`, making ts-arch ignore test files. You could even an artificial `tsconfig.ts-arch.conf` for this
* the second possibility goes in the same direction: `slicesOfProject` could return a builder object which allows changing `tsconfig` properties making the compiler ignore certain files or even do something completely different

Would you mind checking out option one? This would give me some confidence, whether option 2 could be implemented.

Thanks!

**Comment by danielsitnik** [comment]
Thank you for your answer @torfmaster.

I will try the first option, I was not aware that tsarch also ignored the files excluded by tsconfig.
I already have a `tsconfig-build.json` that excludes the test files, will try with that one.

As for the first problem that I reported above, I'm having other issues now because `inFolder` picks up every folder below the one specified (I'ven seen the source code for `folderMatcher`).

For now I was able to work around it by using `matchingPattern` instead with some nasty regex. 😁

**Comment by torfmaster** [comment]
Just to double check: File names are deliberately generated relatively to the specified tsconfig file. If there is no tsconfig explicitly specified it guesses one (by walking up). So could you please check `.inFolder('src/application/service')` instead of `.inFolder('application/service')`?
,
Btw: I am not sure whether checking folders for suffixes really is a good idea here, I would potentially prefer matching the whole file (risking a breaking change).

**Comment by danielsitnik** [comment]
Hey @torfmaster, sorry for the late response, I had some busy days at work.

So I did some tests with a dedicated tsconfig file to exclude the test folder, but it didn't work.
It doesn't matter what I put there on the `exclude` option, my test files are not ignored.

I took the liberty of debugging and taking a look at [extractGraph.ts](https://github.com/ts-arch/ts-arch/blob/32bef1d45b97b45105acc5480d71a5deda929843/src/common/extraction/extractGraph.ts), and from my understanding it doesn't really do anything based on `include` or `exclude` settings from the supplied tsconfig file.

In fact, it resolves a "root dir" (as the documentation says) based on the location of tsconfig, but then it just builds a glob pattern including every ts/tsx file under it, as we can see on [line 78](https://github.com/ts-arch/ts-arch/blob/main/src/common/extraction/extractGraph.ts#L78):
```ts
const rootDir = path.dirname(path.resolve(configFile))
// TODO: make configurable
const globpattern = rootDir + "/**/*.{ts,tsx}"
const files = await guessProjectFiles(globpattern)
```

In my case, since the tsconfig is in the root folder, the result of `guessProjectFiles` is every ts file of my application, even the ones under `node_modules` 😱

I continued to debug and look at the code, and I found nothing to indicate that this list of files is ever manipulated to exclude something.

Please, any thoughts on this?

**Comment by danielsitnik** [comment]
Hey @torfmaster, did you have a chance to look at my last comment?
Thanks.

**Comment by torfmaster** [comment]
Hi @danielsitnik ,

I had busy days, but I finally had a look. Thank you for your patience. I remember me writing these lines because the ts compiler infrastructure didn't pick up all files even if they were included in 'tsconfig.conf`. The downside is that now, too many files are picked up. The question is: what would be a desired solution?

* an "ad hoc" api to ignore certain patterns
* make ts-arch respect the `include` and `exclude` statements

I am a bit in favor of the latter. What do you think?

**Comment by c-harding** [comment]
I think the latter would make much more sense

**Comment by torfmaster** [comment]
Shall I stop asking leading questions? So if @danielsitnik you think this change would be helpful I could start thinking about an implementation.

**Comment by danielsitnik** [comment]
Yes, respecting the include and exclude statements makes more sense.

This way we can have a dedicated `tsconfig-tests.json` with test settings like you mentioned.

Some folders are excluded by default, like: node_modules,bower_components,jspm_packages,outDir

You can find the details on the [tsconfig documentation](https://www.typescriptlang.org/tsconfig#exclude).

**Comment by torfmaster** [comment]
Regarding a possible timeline: I am contributing to `ts-arch` in my spare time so I'll have no time to do this change before Christmas. However, if you from accenture or qaware have capacities, feel free to contribute and I'll be happy to review.

**Comment by torfmaster** [comment]
I just wanted to give you an update: I think I found a pretty straightforward solution and I hope my code will be ready to create a PR soon.
